truex-ios-mobile-reference-app/UnlockSoloViewController.m

```diff
@@ -0,0 +1,313 @@
+    NSLog(@"truex: onAdError: %@", errorMessage);
+    [self resetActiveAdRenderer];
+    [self alertWithTitle:@"onAdError" message:@"Something went wrong, playing fallback ads." completion:nil];
+}
+
+// [4]
+- (void)onNoAdsAvailable {
+    // true[X] - TruexAdRenderer has no ads ready to present, resume with standard ads
+    NSLog(@"truex: onNoAdsAvailable");
+    [self resetActiveAdRenderer];
+    [self alertWithTitle:@"onNoAdsAvailable" message:@"Something went wrong, playing fallback ads." completion:nil];
+}
+
+// [3] - Respond to onAdFreePod
+- (void)onAdFreePod {
+    // true[X] - User has met engagement requirements, skips past remaining pod ads
+    NSLog(@"truex: onAdFreePod");
+    [self.unlocked setOn:YES];
+}
+
+// [5] - Other delegate method
+- (void)onPopupWebsite:(NSString *)url {
+    // true[X] - User wants to open an external link in the true[X] ad
+
+    NSLog(@"truex: onPopupWebsite: %@", url);
+    // Open URL with the SFSafariViewController
+    // SFSafariViewController *svc = [[SFSafariViewController alloc] initWithURL:[NSURL URLWithString: url]];
+    // svc.delegate = self;
+    // svc.modalPresentationStyle = UIModalPresentationOverCurrentContext;
+    // [self presentViewController:svc animated:YES completion:nil];
+    // [self.activeAdRenderer pause];
+    
+    // Or, open the URL directly in Safari
+    // [[UIApplication sharedApplication] openURL:[NSURL URLWithString: url] options:@{} completionHandler:nil];
+    
+    // Or, open with the existing in-app webview
+     UIStoryboard* storyBoard = [UIStoryboard storyboardWithName:@"Main" bundle:nil];
+     WebViewViewController* newViewController = [storyBoard instantiateViewControllerWithIdentifier:@"webviewVC"];
+     newViewController.url = [NSURL URLWithString:url];
+     newViewController.modalPresentationStyle = UIModalPresentationOverCurrentContext;
+     __weak typeof(self) weakSelf = self;
+     newViewController.onDismiss = ^(void) {
+         // true[X] - You will need to pause and remume the true[X] Ad Renderer
+         [weakSelf.activeAdRenderer resume];
+     };
+     [self.activeAdRenderer pause];
+     [self presentViewController:newViewController animated:YES completion:nil];
+}
+
+// When using SFSafariViewController for onPopupWebsite
+//- (void)safariViewControllerDidFinish:(SFSafariViewController *)controller {
+//    // true[X] - You will need remume the true[X] Ad Renderer after safariViewController
+//    if (self.activeAdRenderer) {
+//        [self.activeAdRenderer resume];
+//    }
+//}
+
+
+// MARK: - Helper Functions / Fake Ad Framework
+- (void)fetchAd:(NSString *)url {
+    if (self.vastDictionary != nil) {
+        return;
+    }
+    url = [url stringByReplacingOccurrencesOfString:@"[stream_id]" withString:[[NSUUID UUID] UUIDString]];
+    // replacing user_id with random UUID here for testing, please user the real user ID from the system.
```

Comment:
no, their ad server/app probably has some sort of IDs. Publishers should be sending those instead. 
In this one stage world, the publishers' frontend does not call `get.truex.com` directly anyway. 
The whole IDFA thing is now an option, which is efficiently gone from online research. 

Just putting that comment there so that no one would misunderstand and think-- "trueX needs an random UUID passed into the network_user_id argument."
Which is not the case.